### PR TITLE
RFD: Import EC2 Instance Tags

### PR DESCRIPTION
**Pull request by atburke**
Related: #11627.

### CODE REVIEW

**Review by zmb3 — APPROVED** [review]
Nice and simple. Sounds good to me.

File: rfd/00xx-ec2-tags.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+- Individually adding each tag as a dynamic label
+- Using the AWS API gateway, where the cost scales with the number of nodes using it
+
+As of January 2022, [instance tags are available via the instance metadata service](https://aws.amazon.com/about-aws/whats-new/2022/01/instance-tags-amazon-ec2-instance-metadata-service/). This will allow Teleport nodes to discover their own instance tags. Unlike the AWS API gateway, instance metadata requests are free and per-instance.
+
+## Details
+
+EC2 tags will be supported everywhere that dynamic labels are currently supported (i.e. SSH, Kube, Apps, and Databases).
+
+When a node is created, check if it is running in an EC2 instance. If it is, start a service that periodically (every hour) queries the instance metadata service and updates the tags. Tags created this way will use the `aws` prefix.
```
Comment (hatched) [inline rfd/00xx-ec2-tags.md:29]: I'm assuming that the tags are created on node creation so we should be able to also query the service on startup and create the tags there.

**Review by hatched — APPROVED** [review]

File: rfd/00xx-ec2-tags.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+The current recommended method of [using EC2 tags as Teleport labels](https://goteleport.com/docs/setup/guides/ec2-tags/) requires
+- Creating a custom script to fetch the tags
+- Individually adding each tag as a dynamic label
+- Using the AWS API gateway, where the cost scales with the number of nodes using it
+
+As of January 2022, [instance tags are available via the instance metadata service](https://aws.amazon.com/about-aws/whats-new/2022/01/instance-tags-amazon-ec2-instance-metadata-service/). This will allow Teleport nodes to discover their own instance tags. Unlike the AWS API gateway, instance metadata requests are free and per-instance.
+
+## Details
+
+EC2 tags will be supported everywhere that dynamic labels are currently supported (i.e. SSH, Kube, Apps, and Databases).
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:27]: Could you clarify how this will work for K8s clusters, apps and databases? Currently, a single agent can proxy multiple databases. Will all databases proxies by an agent running on an AWS instance be automatically assigned the instance's labels?

File: rfd/00xx-ec2-tags.md (hunk 1 above)
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:29]: Couple questions here:

> When a node is created,

I would rephrase to "when a Teleport process running an SSH or other service is started".

> Tags created this way will use the `aws` prefix.

Could you clarify what this means? Maybe give a couple examples.

File: rfd/00xx-ec2-tags.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+When a node is created, check if it is running in an EC2 instance. If it is, start a service that periodically (every hour) queries the instance metadata service and updates the tags. Tags created this way will use the `aws` prefix.
+
+In order to use this feature, instance tags in metadata must be enabled for the instance.
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:31]: Can you add details about how this can be enabled and whether it can be enabled for the existing instances?

Comment (r0mant) [reply rfd/00xx-ec2-tags.md:31]: Also, please add details about rate limiting (I know you mentioned it wouldn't be the issue, but would be helpful to mention this in the RFD).

File: rfd/00xx-ec2-tags.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+### Special Tags
+
+If the instance has the tag `Hostname` with a nonempty value, use that value as the node's hostname.
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:35]: Can you add a little bit more details about how this will be implemented? Presumably, the process will check that it's running on AWS when starting up, look for that tag and override whatever is in the config if it's present?

**Review by r0mant — CHANGES_REQUESTED** [review]
Overall makes sense, just asking to flesh out a few more details.

File: rfd/00xx-ec2-tags.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+## Why
+
+The current recommended method of [using EC2 tags as Teleport labels](https://goteleport.com/docs/setup/guides/ec2-tags/) requires
+- Creating a custom script to fetch the tags
+- Individually adding each tag as a dynamic label
+- Using the AWS API gateway, where the cost scales with the number of nodes using it
+
+As of January 2022, [instance tags are available via the instance metadata service](https://aws.amazon.com/about-aws/whats-new/2022/01/instance-tags-amazon-ec2-instance-metadata-service/). This will allow Teleport nodes to discover their own instance tags. Unlike the AWS API gateway, instance metadata requests are free and per-instance.
+
+## Details
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:25]: Can you also add details about how labels reconciliation will work. For example, what happens if you have statically defined labels also? Can they conflict with AWS tags? How will Teleport reconcile AWS labels between runs (i.e. will it detect that a tag was removed from an instance, for example)?

**Review by r0mant — CHANGES_REQUESTED** [review]

**Comment by atburke** [comment]
@r0mant Can you take another look at this?

File: rfd/00xx-ec2-tags.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+In order to use this feature, instance tags in metadata must be enabled for the instance. Instance tags in metadata can be enabled/disabled when launching a new instance; they can also be toggled for an existing instance via `Actions > Instance settings > Allow tags in instance metadata` in the management console or with the `modify-instance-metadata-options` command in the AWS CLI. See the [AWS documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html#allow-access-to-tags-in-IMDS) for more details.
+
+### Throttling
+AWS applies [per-instance throttling to instance metadata requests](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instancedata-throttling). Each instance can have at most 50 tags, so with one request to fetch all tag keys and up to 50 requests to fetch tag values, Teleport only needs to make at most 51 instance metadata requests per hour, which should not have throttling issues.
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:36]: Oh so you have to make an additional request per tag to retrieve its value? That's a bit of a bummer but I guess it's fine since that's how it works.

File: rfd/00xx-ec2-tags.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+### Special Tags
+
+#### `Hostname`
+
+When a Teleport process is created, it will check if it is running in an EC2 instance. If it is, and the instance has the tag `Hostname` with a nonempty value, the process will use that value as the node's hostname, overriding the hostname provided in the config.
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:42]: I would probably choose some tag name that is less likely to already exist on the instances so avoid unexpectedly changing hostname for existing clusters. How about `TeleportHostname`?

File: rfd/00xx-ec2-tags.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+EC2 labels will use the `aws` prefix in Teleport for namespacing (similar to `teleport.dev`). For example, if the instance has the tag `Name: my-instance`, Teleport will add the label `aws/Name: my-instance`. Ideally, the `aws` namespace will prevent collisions with static and command label names, but if there is a collision, EC2 labels will not override static or command labels.
+
+In order to use this feature, instance tags in metadata must be enabled for the instance. Instance tags in metadata can be enabled/disabled when launching a new instance; they can also be toggled for an existing instance via `Actions > Instance settings > Allow tags in instance metadata` in the management console or with the `modify-instance-metadata-options` command in the AWS CLI. See the [AWS documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html#allow-access-to-tags-in-IMDS) for more details.
```
Comment (r0mant) [inline rfd/00xx-ec2-tags.md:33]: I was reading [Working with instance tags in instance metadata](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html#work-with-tags-in-IMDS) and stumbled upon the following paragraph:

> If you add or remove an instance tag, the instance metadata is updated while the instance is running for [instances built on the Nitro System](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#ec2-nitro-instances), without needing to stop and then start the instance. For all other instances, to update the tags in the instance metadata, you must stop and then start the instance.

Does this mean that for older instance types metadata will not get updated tags until the instance is restarted?

Comment (r0mant) [reply rfd/00xx-ec2-tags.md:33]: It may be fine as it seems like most recent instance types like m5, c5, etc. are all built on Nitro system but it might be worth making sure and calling this out explicitly in the RFD and in the docs we will write for this feature.

**Review by r0mant — APPROVED** [review]
lgtm with the latest round of comments
